### PR TITLE
a few ideas

### PR DESCRIPTION
**Pull request by oisdk**
Maybe not so much error checking? If you check for errors within each function, that'll create a pretty big overhead (especially for, say functions like your single random number: if you check that `max > min` each time for that, about a quarter of the processor cycles could be spent on that check). However, if you remove the checks, and specify that max must be greater than min in the docs (this is how the standard library `Range` does it) then you allow the user to do their own error checking, or bypass it if they know that the arguments they're using will always work, saving them the overhead.

Also, optionals usually aren't used for error checking: they're usually used for things that can and do return `nil` in normal operations. Look at your array sampling functions, and checking if the length parameter is less than zero. Realistically, that's something that should never happen, no matter what. You don't really want your function to return nil in that case, you want it to crash, so you find the bug quickly and early.

The standard library `find()` function is a good example of something that returns an optional, instead of crashing. `find()` returns an optional because of the nature of what it does: if it doesn't find what you're looking for, it returns `nil`. This may be something that you want to happen. However, if there was some other error, some problem in the logic of what it was doing (say you passed it the wrong type), you don't want it to return `nil`, you want it to crash. The difference is that you want your function to crash for arguments it should _never_ get, but return `nil` for a failure that might be desirable.

Some of the errors you check for don't actually need to be checked: say the sampling with replacement function. Why does length have to be smaller than the array you're sampling from? Say I want to sample `3` from the array `[1, 2]`. Why can't you just return, say `[1, 1, 2]`? It makes sense for the function to do that. (However, It wouldn't make sense for the `samplingWithoutReplacement` function to allow this, for example) Also, your `samplingWithGivenProbabilities` function doesn't need its probabilities to add up to one: you can just total its probabilities, and use that as the factor for your random number generator from then on in.

I think your variable names are a little long, and your functions take too many named parameters. Making the names long makes it annoying to type, and it can take up a lot of space, making the code less readable. The extra information given in the names often isn't needed: look at `randomContinuousUniformArray` for example. Why does that have the word `Array` at the end? You know it's going to return an array - it's right there in the type signature. Same goes for `samplingWithReplacementFromArray`: you know that it's got to be `FromArray` without having it in the name, because it only accepts an Array as its argument. In fact, in a lot of places, you probably don't have to specify named parameters at all: since your docs are so thorough, every piece of information you need is just an alt-click away. Especially in the one-parameter functions.

You seem to do a lot of initialising arrays to `0.0`, and then looping through their count, and filling them. The functions like `map()`, `flatMap()`, etc., are more efficient: instead of doing initialise, loop, lookup, change, increment, you can do it in one.

I wouldn't use `0`s and `1`s in place of `Bool`s, if at all possible. It obfuscates the purpose, it's slower, and you're likely going to have to convert back to a `Bool` at some stage anyway.

I hope I didn't come across as negative:  I thought it was really cool. I had a lot of fun looking through it!  Most of the things I'm talking about here are taste-based, rather than hard and fast rules, anyway, so don't feel any obligation to change what you've written: if you disagree with anything I've said, ignore it! I'm no authority.

Thanks for the cool code!

### CODE REVIEW

**Comment by SebastianOsinski** [comment]
Hi. The first thing: I didn't expect such vast feedback from anyone, when I created this repository. Thank you! Those are my first steps in Swift programming and I don't have big programming background, so this means a lot to me :). 

Thank you for the explanation about using Optionals. I misunderstood the whole purpose behind them.

To be honest, this length checking in sampling with replacement was copy-paste from other function, which I didn't change. 

Regarding the names of functions, I agree with your remarks, but I would like to make it my own way. I will surely remove those redundant `fromArray` etc.

Thank you for pointing out the existence of `map()` and other useful functions, I will use them, but I want to do it myself, purely for educational value :) 

So, to sum up, I really thank you for your time and remarks and I will surely incorporate some changes you proposed.
